projects/RabbitMQ.Client/client/TaskExtensions.cs

```diff
@@ -59,21 +59,29 @@ public static Task TimeoutAfter(this Task task, TimeSpan timeout)
 
             return DoTimeoutAfter(task, timeout);
 
+            // https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#using-a-timeout
             static async Task DoTimeoutAfter(Task task, TimeSpan timeout)
             {
-                if (task == await Task.WhenAny(task, Task.Delay(timeout)).ConfigureAwait(false))
+                using (var cts = new CancellationTokenSource())
                 {
-                    await task.ConfigureAwait(false);
-                }
-                else
-                {
-                    Task supressErrorTask = task.ContinueWith(
-                        continuationAction: continuation,
-                        state: null,
-                        cancellationToken: CancellationToken.None,
-                        continuationOptions: s_tco,
-                        scheduler: TaskScheduler.Default);
-                    throw new TimeoutException();
+                    Task delayTask = Task.Delay(timeout, cts.Token);
+                    Task resultTask = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
+                    if (resultTask == delayTask)
+                    {
+                        Task supressErrorTask = task.ContinueWith(
+                            continuationAction: continuation,
+                            state: null,
+                            cancellationToken: CancellationToken.None,
+                            continuationOptions: s_tco,
+                            scheduler: TaskScheduler.Default);
+                        throw new TimeoutException();
+                    }
+                    else
+                    {
+                        cts.Cancel();
+                    }
+
+                    await task;
```

Comment:
.ConfigureAwait(false)

Reply:
Every `await` should be paired with a `ConfigureAwait(false)`.

Reply:
@paulomorgado yes I realize that, I just missed this one. Once I get the analyzer in place hopefully this will be less of an issue.